pkg/generator/scheduler.go

```diff
@@ -0,0 +1,151 @@
+// protocol is executing or not. This interface should be implemented by all
+// important protocols of the client, such as distributed key generation or
+// signing.
+type Protocol interface {
+	IsExecuting() bool
+}
+
+// Scheduler allows managing computationally heavy operations: stopping and
+// resuming them. The client needs to generate parameters for cryptographic
+// algorithms and generating these parameters requires a lot of CPU cycles.
+// The generation process may starve other processes in the client when it comes
+// to access to the CPU. The scheduler allows starting the parameter generation
+// when no other processes such as key generation or signing are executed on the
+// client. This way, the client that would normally be idle, can spend CPU
+// cycles on computationally heavy operations and stop these operations when CPU
+// cycles are needed elsewhere.
+type Scheduler struct {
+	state     state
+	workers   []func(context.Context)
+	stops     []context.CancelFunc
+	workMutex sync.Mutex
+
+	protocols      []Protocol
+	protocolsMutex sync.Mutex
+}
+
+func (s *Scheduler) RegisterProtocol(protocol Protocol) {
+	s.protocolsMutex.Lock()
+	defer s.protocolsMutex.Unlock()
+
+	s.protocols = append(s.protocols, protocol)
+}
+
+// Compute takes the worker function and starts the computations in a separate
+// goroutine if the scheduler status is "working". Otherwise, when the scheduler
+// status is "stopped", the worker function is scheduled for execution later.
+// The function accepts the context and is required to stop the execution if
+// the context is done. The function will be called in a loop until the
+// scheduler is stopped.
+func (s *Scheduler) compute(workerFn func(context.Context)) {
+	s.workMutex.Lock()
+	defer s.workMutex.Unlock()
+
+	s.workers = append(s.workers, workerFn)
+
+	if s.state == working {
+		s.startWorker(workerFn)
+	}
+}
+
+// Stop asks all worker functions to stop their work. The context passed to
+// the function is cancelled and no further calls to the worker function are
+// done until the scheduler work is resumed.
+func (s *Scheduler) stop() {
+	s.workMutex.Lock()
+	defer s.workMutex.Unlock()
+
+	if s.state == stopped {
+		return
+	}
+
+	logger.Info("stopping computations\n")
+	s.state = stopped
+
+	for _, stop := range s.stops {
+		stop()
+	}
+	s.stops = nil
+}
+
+// Resume resumes the work of all worker functions, each in a separate
+// goroutine.
+func (s *Scheduler) resume() {
+	s.workMutex.Lock()
+	defer s.workMutex.Unlock()
+
+	if s.state == working {
+		return
+	}
+
+	logger.Info("resuming computations\n")
+	s.state = working
+
+	for _, worker := range s.workers {
+		s.startWorker(worker)
+	}
+}
+
+// StartWorker takes the provided worker function, creates for it an individual
+// context and starts executing it in the loop until the context is done.
+func (s *Scheduler) startWorker(workerFn func(context.Context)) {
```

Comment:
Can we add a mention that this function should be executed only under the condition that the `workMutex` is locked?

Reply:
I think `workMutex` has nothing to do here. This function should only be called when the scheduler is working but most importantly, it should only be called by the scheduler. I added this info in 91fc27a77c9976bf147aa942a2d08dd9940c267d but I did not mention the mutex since I do not follow the comment. Can you explain it, please?

Reply:
The `startWorker` function is modifying the `s.stops` slice that is protected by the `s.workMutex`. That said, the `startWorker` call should be made only if the `s.workMutex` is locked. Currently, `startWorker` is called by `resume` and `compute` that already lock the  `workMutex` so `startWorker` cannot do it on their own as that would cause a deadlock. My comment was about mentioning the fact that `startWorker` does not lock the mutex explicitly but has an implicit requirement that it must be called on the locked mutex.

Reply:
Ah, right! 03539e534ec6b10ac9ea899ec820464822c528eb